### ISSUE TITLE
Clever Edits for Insane Tree-Sitter Performance

### ISSUE THREAD

**Issue by theHamsta** [steelsojka/pears.nvim#11]
I'm often having performance problems with tree-sitter for C++ and Javascript on large files. The issue seems to be that typing `{` seems to cause a re-parse often until the end of the file which does not seem to happen for most other parsers.

I was wondering whether it would be possible to perform `{}` as a single edit as compared to `{` and then expand `}`. Pushing `{}` as one edit should really cause the tree to stay the same for single byte pairs.

I know this is the wrong way to face a possible performance problem. But I wound find it extremely elegant when parent matching also helps to always have valid syntax trees.

**Comment by theHamsta** [comment]
Maybe we could allow to defer parsing on certain characters and wait for the next `onbytes` or remap `{` to some `maybe_expand`

**Comment by steelsojka** [comment]
I've thought about just using `v:chars` to insert the text instead of running the commands to perform the expansion, but it ran into some problems. Indentation doesn't seem to work right if the full expansion is inserted instead of doing it in a scheduled vim callback. This would only work for simple pairs that don't have further expansions. For example the pairs `<` and `>` would expand fine but if we have a pair of `<!--` and `-->` we delete the previous expanded pair and insert the new one, which we wouldn't be able to do with replacing the `v:char`. I can look into it though. Did you have performance issues with other pair plugins?

**Comment by theHamsta** [comment]
All other pair plugins where far to annoying. So I didn't use them. pears.nvim is the first that I trust!

To be clear the performance issue is not the pair plugin. It's actually without the plugin. I thought it would be cool if the languagetree could hold back its parsing until the closing pair is inserted which is a totally different edit. It would be basically skipping the invalid states of the parsed tree by after the pears.nvim has restored a syntactically valid document. This is actually a bit against the goals of tree since it's for parsing also invalids documents so this was just a random idea.

The *issues* are actually not really a problem. At worst just a short flickering.

**Comment by theHamsta** [comment]
We had the issue basically here: https://github.com/p00f/nvim-ts-rainbow/pull/29 Incremental updates don't have too much of an advantage because for C++/JS when you open a `{` it is basically a changedtree for the whole document. For most other parsers this is still localized in some way.

**Comment by steelsojka** [comment]
Oh I see. So basically kind of disabling the treesitter parsing until we have done all edits for the pair?

**Comment by steelsojka** [comment]
I believe, right now it inserts the char normally and then expands on a vim scheduled function. We could just set `v:char` to an empty string and handle the character insertion ourselves instead in the scheduled function... That may only trigger it once. I to have noticed some flickering.

**Comment by theHamsta** [comment]
> Oh I see. So basically kind of disabling the treesitter parsing until we have done all edits for the pair?

yes, this is what I meant. But probably not worth the effort. It only happens on extremely big files with a lot of query regexing and logic.



https://user-images.githubusercontent.com/7189118/115715405-602e9280-a378-11eb-90a9-222374302e91.mp4

The flickering basically comes because the document has a completely different structure for a second. When you edit manually this also happens.
https://user-images.githubusercontent.com/7189118/115716094-1e521c00-a379-11eb-8097-0a44a0870350.mp4

But again this is not really an issue of this plugin and I should probably close this issue.

**Comment by steelsojka** [comment]
I do notice some slow down when inserting a pair via v:char and then expanding vs doing it all in a scheduled function. I'll take a look at it.

**Comment by steelsojka** [comment]
@theHamsta Can you try this branch https://github.com/steelsojka/pears.nvim/tree/fix/single-edit and see if there are any issues? This will insert the character and expand the pair within the same schedule callback.

**Comment by steelsojka** [comment]
@theHamsta I accidentally pushed some refactored code on that branch, so..... it may not work. I'll let you know when I have something stable.

**Comment by steelsojka** [comment]
@theHamsta I merge my changes into master. Let me know if that speeds up those flickers. In my limited testing it seemed to speed up placing "(" ")" pairs in a large JS file.

**Comment by theHamsta** [comment]
Yes it seems a lot lot smoother. No flickering at all.

However one of the changes of today causes the deletion of some text for me. 

https://user-images.githubusercontent.com/7189118/115886349-1fef1300-a451-11eb-9930-0c1128a0951d.mp4

**Comment by theHamsta** [comment]
Next GPT3 language model detects when you write bad code and deletes it immediatelly.

**Comment by theHamsta** [comment]
It's related with my usage of compe. With autocomplete disabled it won't delete any text.

**Comment by theHamsta** [comment]
Did some changes in my compe config and the problem does no longer occur.

**Comment by theHamsta** [comment]
As compe did this before it is probably compe to blame. Will check whether o can reproduce with other completion plugins

**Comment by steelsojka** [comment]
Hmmm weird. I did some stuff with multibyte pairs that may have effected the deletion. I'll take a look and see if I can reproduce it.

**Comment by steelsojka** [comment]
Can you give me a run down of what's happening?

**Comment by theHamsta** [comment]
- type foo
- autocomplete shows up
- type (
- text gets deleted

I guess a hook of compe expects ( to appear after the key press I'll have a look at their code base soon or it's something related to insertcomplete mode

I can prevent the deletion

- type foo
- autocomplete appears
- select entry and press c-y complete mode ends
- type (
- () appears

**Comment by theHamsta** [comment]
Also `compe#close('<C-e>')` get pasted as because `()` and `''` expand together

![image](https://user-images.githubusercontent.com/7189118/115918890-62781600-a478-11eb-8af6-6ff7f1c12874.png)

**Comment by nyngwang** [comment]
I just want to know when the performance problem will be resolved.
